src/plugins/kibana/public/settings/sections/indices/add_data_steps/pipeline_setup/lib/pipeline.js

```diff
@@ -0,0 +1,158 @@
+import _ from 'lodash';
+
+export default function Pipeline(processorTypes, Processor) {
+  const self = this;
+
+  self.processors = [];
+  self.counter = 0;
+  self.input = {};
+  self.output = undefined;
+  self.dirty = false;
+  self.Processor = Processor;
+  self.processorTypes = processorTypes;
+};
+
+Pipeline.prototype.load = function (pipeline) {
+  const self = this;
+
+  while (self.processors.length > 0) {
+    self.processors.pop();
+  }
+
+  pipeline.processors.forEach((processor) => {
+    self.addExisting(processor);
+  });
+};
+
+Pipeline.prototype.remove = function (processor) {
+  const self = this;
+  const processors = self.processors;
+  const index = processors.indexOf(processor);
+
+  processors.splice(index, 1);
+};
+
+Pipeline.prototype.moveUp = function (processor) {
+  const self = this;
+  const processors = self.processors;
+  const index = processors.indexOf(processor);
+
+  if (index === 0) return;
+
+  const temp = processors[index - 1];
+  processors[index - 1] = processors[index];
+  processors[index] = temp;
+};
+
+Pipeline.prototype.moveDown = function (processor) {
+  const self = this;
+  const processors = self.processors;
+  const index = processors.indexOf(processor);
+
+  if (index === processors.length - 1) return;
+
+  const temp = processors[index + 1];
+  processors[index + 1] = processors[index];
+  processors[index] = temp;
+};
+
+function remove(array, arrayToRemove) {
+  arrayToRemove.forEach((itemToRemove) => {
+    _.remove(array, (o) => { return o === itemToRemove; });
+  });
+};
+
+Pipeline.prototype.addExisting = function (existingProcessor) {
+  const self = this;
+  const processors = self.processors;
+
+  const processorType = _.find(self.processorTypes, (o) => { return o.typeId === existingProcessor.typeId; });
+  const newProcessor = self.add(processorType);
+
+  const keys = _.keys(existingProcessor);
+  remove(keys, ['title', 'template', 'typeId', 'processorId', 'outputObject', 'inputObject', 'description']);
+  keys.forEach((key) => {
+    _.set(newProcessor, key, _.get(existingProcessor, key));
+  });
```

Comment:
This could be simplified:

```
_.assign(newProcessor, _.omit(existingProcessor, ['title', 'template', 'typeId', 'processorId', 'outputObject', 'inputObject', 'description']));
```